lib/ContentTypeParser.js

```diff
@@ -1,67 +1,181 @@
+
 function ContentTypeParser () {
   this.customParsers = {}
-  this.parserList = []
+  this.customParsers['application/json'] = defaultJsonParser
+  this.parserList = ['application/json']
+  this.cache = lru(100)
 }
 
-ContentTypeParser.prototype.add = function (contentType, fn) {
+ContentTypeParser.prototype.add = function (contentType, opts, parser) {
   if (typeof contentType !== 'string') throw new Error('The content type should be a string')
   if (contentType.length === 0) throw new Error('The content type cannot be an empty string')
-  if (typeof fn !== 'function') throw new Error('The content type handler should be a function')
+  if (typeof parser !== 'function') throw new Error('The content type handler should be a function')
 
   if (this.hasParser(contentType)) {
     throw new Error(`Content type parser '${contentType}' already present.`)
   }
+
+  parser.asString = opts.asString === true
+  parser.asBuffer = opts.asBuffer === true
+
   if (contentType === '*') {
     this.parserList.push('')
-    this.customParsers[''] = fn
+    this.customParsers[''] = parser
   } else {
-    this.parserList.unshift(contentType)
-    this.customParsers[contentType] = fn
+    if (contentType !== 'application/json') {
+      this.parserList.unshift(contentType)
+    }
+    this.customParsers[contentType] = parser
   }
 }
 
 ContentTypeParser.prototype.hasParser = function (contentType) {
+  if (contentType === 'application/json') return false
   return contentType in this.customParsers
 }
 
 ContentTypeParser.prototype.fastHasHeader = function (contentType) {
-  if (!contentType) return this.hasParser('')
-
   for (var i = 0; i < this.parserList.length; i++) {
     if (contentType.indexOf(this.parserList[i]) > -1) return true
   }
   return false
 }
 
-ContentTypeParser.prototype.getHandler = function (contentType) {
-  if (contentType === undefined) return this.customParsers['']
-
+ContentTypeParser.prototype.getParser = function (contentType) {
   for (var i = 0; i < this.parserList.length; i++) {
     if (contentType.indexOf(this.parserList[i]) > -1) {
-      return this.customParsers[this.parserList[i]]
+      var parser = this.customParsers[this.parserList[i]]
+      this.cache.set(contentType, parser)
+      return parser
     }
   }
+
+  return this.customParsers['']
 }
 
 ContentTypeParser.prototype.run = function (contentType, handler, request, reply) {
-  var result = this.getHandler(contentType)(request.raw, done)
-  if (result && typeof result.then === 'function') {
-    result
-      .then(body => done(null, body))
-      .catch(done)
+  var parser = this.cache.get[contentType] || this.getParser(contentType)
+  if (parser === undefined) {
+    reply.code(415).send(new Error('Unsupported Media Type: ' + contentType))
+  } else {
+    if (parser.asString === true || parser.asBuffer === true) {
+      rawBody(
+        request,
+        reply,
+        reply.context._parserOptions,
+        parser,
+        handler
```

Comment:
I think you can pass `done` here instead of `handler`. That way the `done` function won't need to be duplicated (line 159).